fem/field_interpolant.hpp

```diff
@@ -0,0 +1,188 @@
+#include "gridfunc.hpp"
+#ifdef MFEM_USE_MPI
+#include "pgridfunc.hpp"
+#include "pbilinearform.hpp"
+#endif
+
+namespace mfem
+{
+
+class FieldInterpolant
+{
+protected:
+   bool setup_disc;
+   bool setup_full;
+   Vector m_all_data;
+   BilinearForm *L2;
+   CGSolver *cg;
+   int NE;
+public:
+   // The FiniteElementSpace passed into here should have a vdim set to 1 in order for the
+   // MassIntegrator to work properly if the ProjectQuadratureCoefficient method is used with
+   // a VectorQuadratureFunctionCoefficient.
+   FieldInterpolant(FiniteElementSpace *fes) : setup_disc(false), setup_full(false)
+   {
+      L2 = new BilinearForm(fes);
+
+      const FiniteElement &el = *fes->GetFE(0);
+      const IntegrationRule *ir = &(IntRules.Get(el.GetGeomType(),
+                                                 2 * el.GetOrder() + 1));
+
+      L2->AddDomainIntegrator(new MassIntegrator(ir));
+      L2->Assemble();
+   }
+   // This function takes a vector quadrature function coefficient and projects it onto a GridFunction that lives
+   // in L2 space. This function requires fes to be L2 finite element space that we're projecting onto.
+   void ProjectQuadratureDiscCoefficient(GridFunction &gf,
+                                         VectorQuadratureFunctionCoefficient &vqfc,
+                                         FiniteElementSpace &fes);
+   // This function takes a quadrature function coefficient and projects it onto a GridFunction that lives
+   // in L2 space. This function requires fes to be L2 finite element space that we're projecting onto.
+   void ProjectQuadratureDiscCoefficient(GridFunction &gf,
+                                         QuadratureFunctionCoefficient &qfc,
+                                         FiniteElementSpace &fes);
+   // This function takes a vector quadrature function coefficient and projects it onto a GridFunction of the same space as vector
+   // quadrature function coefficient.
+   void ProjectQuadratureCoefficient(GridFunction &gf,
+                                     VectorQuadratureFunctionCoefficient &vqfc,
+                                     FiniteElementSpace &fes);
+   // This function takes a quadrature function coefficient and projects it onto a GridFunction of the same space as
+   // quadrature function coefficient.
+   void ProjectQuadratureCoefficient(GridFunction &gf,
+                                     QuadratureFunctionCoefficient &qfc,
+                                     FiniteElementSpace &fes);
+   // Tells the ProjectQuadratureDiscCoefficient that they need to recalculate the data.
+   void SetupDiscReset() { setup_disc = false; }
+   // Tells the ProjectQuadratureCoefficient that they need to recalculate the data.
+   virtual void FullReset()
+   {
+      L2->Update();
+      L2->Assemble();
+   }
+   virtual void SetupCG()
+   {
+      cg = new CGSolver();
+      cg->SetPrintLevel(0);
+      cg->SetMaxIter(2000);
+      cg->SetRelTol(sqrt(1e-30));
```

Comment:
sqrt(1-e30) = 1e-15

Reply:
Fixed and also the user can now specify all of these values, but if they don't default values are provided.